File: node/src/carrier.ts
import { HttpClient } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface Carrier {
    id: number;
    name: string;
    code: string;
    carrier_type: string;
    tracking_type: string;
    tracking_provider?: string;
    tracking_slug?: string;
    business_type: string;
    state: 'active' | 'pending' | 'suspended';
    description: string;
    website: string;
    contact_email?: string;
    contact_phone?: string;
    created_at: string;
    updated_at: string;
}

// ============ Carrier Client ============

export class CarrierClient extends HttpClient {
    async list(filter: { page?: number; page_size?: number; state?: string } = {}): Promise<ApiResult<Carrier[]>> {
        const q = this.buildQuery(filter);
        const result = await this.request(`/api/v1/carrier/list${q}`) as any;
        return result as ApiResult<Carrier[]>;
    }

    async get(id: string): Promise<ApiResult<Carrier>> {
        const result = await this.request(`/api/v1/carrier/${encodeURIComponent(id)}`) as any;
        return result as ApiResult<Carrier>;
    }

    async create(data: { name: string; code: string; carrier_type?: string; tracking_type?: string; business_type?: string }): Promise<ApiResult<Carrier>> {
        const result = await this.request('/api/v1/carrier/register', 'POST', data) as any;
        return result as ApiResult<Carrier>;
    }

    async update(id: string, data: Partial<Carrier>): Promise<ApiResult<Carrier>> {
        const result = await this.request(`/api/v1/carrier/${encodeURIComponent(id)}`, 'PUT', data) as any;
        return result as ApiResult<Carrier>;
    }

    async delete(id: string): Promise<ApiResult<{ deleted: boolean }>> {
        const result = await this.request(`/api/v1/carrier/${encodeURIComponent(id)}`, 'DELETE') as any;
        return result as ApiResult<{ deleted: boolean }>;
    }
}

File: rust/src/carrier.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;
use crate::carrier_types::Carrier;

pub struct CarrierClient {
    pub(crate) inner: HttpClient,
}

impl CarrierClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }

    pub async fn list(
        &self,
        page: i32,
        page_size: i32,
        state: Option<&str>,
    ) -> Result<serde_json::Value> {
        let mut params = vec![
            format!("page={}", page),
            format!("page_size={}", page_size),
        ];
        if let Some(s) = state {
            params.push(format!("state={}", urlencoding::encode(s)));
        }
        let path = format!("/api/v1/carrier/list?{}", params.join("&"));
        self.inner.request(&path, Method::GET, None).await
    }

    pub async fn get(&self, id: &str) -> Result<Carrier> {
        self.inner
            .request(
                &format!("/api/v1/carrier/{}", id),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<Carrier> {
        self.inner
            .request("/api/v1/carrier/register", Method::POST, Some(body))
            .await
    }

    pub async fn update(&self, id: &str, body: serde_json::Value) -> Result<Carrier> {
        self.inner
            .request(
                &format!("/api/v1/carrier/{}", id),
                Method::PUT,
                Some(body),
            )
            .await
    }

    pub async fn delete(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/carrier/{}", id),
                Method::DELETE,
                None,
            )
            .await
    }
}
